// react-images/lib/components/Spinner.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _noImportant = require("aphrodite/no-important");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var Spinner = function Spinner(props) {
  var classes = _noImportant.StyleSheet.create(styles(props));

  return _react["default"].createElement("div", {
    className: (0, _noImportant.css)(classes.spinner)
  }, _react["default"].createElement("div", {
    className: (0, _noImportant.css)(classes.ripple)
  }));
};

Spinner.propTypes = {
  color: _propTypes["default"].string,
  size: _propTypes["default"].number
};
var rippleKeyframes = {
  '0%': {
    top: '50%',
    left: '50%',
    width: 0,
    height: 0,
    opacity: 1
  },
  '100%': {
    top: 0,
    left: 0,
    width: '100%',
    height: '100%',
    opacity: 0
  }
};

var styles = function styles(_ref) {
  var color = _ref.color,
      size = _ref.size;
  return {
    spinner: {
      display: 'inline-block',
      position: 'relative',
      width: size,
      height: size
    },
    ripple: {
      position: 'absolute',
      border: "4px solid ".concat(color),
      opacity: 1,
      borderRadius: '50%',
      animationName: rippleKeyframes,
      animationDuration: '1s',
      animationTimingFunction: 'cubic-bezier(0, 0.2, 0.8, 1)',
      animationIterationCount: 'infinite'
    }
  };
};

var _default = Spinner;
exports["default"] = _default;
