// react-images/lib/components/View.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.viewCSS = void 0;

var _react = _interopRequireDefault(require("react"));

var _glam = _interopRequireDefault(require("glam"));

var _primitives = require("../primitives");

var _utils = require("../utils");

var _componentHelpers = require("./component-helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var viewCSS = function viewCSS() {
  return {
    lineHeight: 0,
    position: 'relative',
    textAlign: 'center'
  };
};

exports.viewCSS = viewCSS;

var View = function View(props) {
  var data = props.data,
      formatters = props.formatters,
      getStyles = props.getStyles,
      index = props.index,
      isFullscreen = props.isFullscreen,
      isModal = props.isModal;
  var innerProps = {
    alt: formatters.getAltText({
      data: data,
      index: index
    }),
    src: (0, _componentHelpers.getSource)({
      data: data,
      isFullscreen: isFullscreen
    })
  };
  return (0, _glam["default"])(_primitives.Div, {
    css: getStyles('view', props),
    className: (0, _utils.className)('view', {
      isFullscreen: isFullscreen,
      isModal: isModal
    })
  }, (0, _glam["default"])(_primitives.Img, _extends({}, innerProps, {
    className: (0, _utils.className)('view-image', {
      isFullscreen: isFullscreen,
      isModal: isModal
    }),
    css: {
      height: 'auto',
      maxHeight: '100vh',
      maxWidth: '100vw',
      userSelect: 'none'
    }
  })));
};

var _default = View;
exports["default"] = _default;
