// react-images/lib/formatters.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

// NOTE: props aren't used by default for some getters but consumers may need
// them, this needs to be reflected in the flow type.

/* eslint-disable no-unused-vars */
// ==============================
// Navigation
// ==============================

/* ARIA label for the next button */
function getNextLabel(_ref) {
  var currentIndex = _ref.currentIndex,
      views = _ref.views;
  return "Show slide ".concat(currentIndex + 2, " of ").concat(views.length);
}
/* ARIA label for the previous button */


function getPrevLabel(_ref2) {
  var currentIndex = _ref2.currentIndex,
      views = _ref2.views;
  return "Show slide ".concat(currentIndex, " of ").concat(views.length);
}
/* HTML title for the next button */


function getNextTitle(props) {
  return 'Next (right arrow)';
}
/* HTML title for the previous button */


function getPrevTitle(props) {
  return 'Previous (left arrow)';
} // ==============================
// Header
// ==============================


/* ARIA label for the close button */
function getCloseLabel(props) {
  return 'Close (esc)';
}
/* ARIA label for the fullscreen button */


function getFullscreenLabel(_ref3) {
  var isFullscreen = _ref3.isFullscreen;
  return isFullscreen ? 'Exit fullscreen (f)' : 'Enter fullscreen (f)';
} // ==============================
// View
// ==============================

/* alt text for each image in the carousel */


function getAltText(_ref4) {
  var data = _ref4.data,
      index = _ref4.index;
  if (data.caption) return data.caption;
  return "Image ".concat(index + 1);
} // ==============================
// Exports
// ==============================


var _default = {
  getAltText: getAltText,
  getNextLabel: getNextLabel,
  getPrevLabel: getPrevLabel,
  getNextTitle: getNextTitle,
  getPrevTitle: getPrevTitle,
  getCloseLabel: getCloseLabel,
  getFullscreenLabel: getFullscreenLabel
};
exports["default"] = _default;
